lib/active_merchant/billing/gateways/worldpay.rb

```diff
@@ -321,9 +325,24 @@ def parse_element(raw, node)
         raw
       end
 
-      def commit(action, request, *success_criteria)
-        xmr = ssl_post(url, request, 'Content-Type' => 'text/xml', 'Authorization' => encoded_credentials)
+      def headers(options)
+        headers = {
+          'Content-Type' => 'text/xml',
+          'Authorization' => encoded_credentials
+        }
+        if options[:cookie]
+          headers.merge!('Set-Cookie' => options[:cookie])
+        end
+        headers
+      end
+
+      def commit(action, request, *success_criteria, options)
+        xmr = ssl_post(url, request, headers(options))
         raw = parse(action, xmr)
+        if options[:execute_threed]
+          raw[:cookie] = @cookie
+          raw[:session_id] = options[:session_id]
```

Comment:
so the presence of the cookie and session are the only things flagging a request to execute 3DS? Weird.

Reply:
@curiousepic the `acceptHeader` and `userAgentHeader` must also be sent, but I didn't encapsulate that in a `if`